sdk/core/Azure.Core.Experimental/tests/LowLevelClientTests.cs

```diff
@@ -211,6 +212,145 @@ await client.GetPetAsync("snoopy", new RequestContext()
+
+            var context = new RequestContext();
+            context.AddPolicy(new AddHeaderPolicy("PerCallHeader", "Value"), HttpPipelinePosition.PerCall);
+
+            Response response = await client.GetPetAsync("snoopy", context);
+
+            Request request = mockTransport.Requests[0];
+            Assert.IsTrue(request.Headers.TryGetValues("PerCallHeader", out var values));
+            Assert.AreEqual(1, values.Count());
+            Assert.AreEqual("Value", values.ElementAt(0));
+        }
+
+        [Test]
+        public async Task CanAddPerRetryPolicy()
+        {
+            var retryResponse = new MockResponse(408); // Request Timeout
+            var petResponse = new MockResponse(200);
+
+            Pet pet = new("snoopy", "beagle");
+            petResponse.SetContent(SerializationHelpers.Serialize(pet, SerializePet));
+
+            // retry twice -- this will add the header three times.
+            var mockTransport = new MockTransport(retryResponse, retryResponse, petResponse);
+            PetStoreClient client = CreateClient(mockTransport);
+
+            var context = new RequestContext();
+            context.AddPolicy(new AddHeaderPolicy("PerRetryHeader", "Value"), HttpPipelinePosition.PerRetry);
+
+            Response response = await client.GetPetAsync("snoopy", context);
+
+            Request request = mockTransport.Requests[0];
+            Assert.IsTrue(request.Headers.TryGetValues("PerRetryHeader", out var values));
+            Assert.AreEqual(3, values.Count());
+            Assert.AreEqual("Value", values.ElementAt(0));
+            Assert.AreEqual("Value", values.ElementAt(1));
+            Assert.AreEqual("Value", values.ElementAt(2));
+        }
+
+        [Test]
+        public async Task CanAddBeforeTransportPolicy()
+        {
+            var retryResponse = new MockResponse(408); // Request Timeout
+            var petResponse = new MockResponse(200);
+
+            Pet pet = new("snoopy", "beagle");
+            petResponse.SetContent(SerializationHelpers.Serialize(pet, SerializePet));
+
+            // retry twice
+            var mockTransport = new MockTransport(retryResponse, retryResponse, petResponse);
+            PetStoreClient client = CreateClient(mockTransport);
+
+            var context = new RequestContext();
+            context.AddPolicy(new AddHeaderPolicy("BeforeTransportHeader", "Value"), HttpPipelinePosition.BeforeTransport);
+
+            Response response = await client.GetPetAsync("snoopy", context);
+
+            Request request = mockTransport.Requests[0];
+            Assert.IsTrue(request.Headers.TryGetValues("BeforeTransportHeader", out var values));
+            Assert.AreEqual(3, values.Count());
+            Assert.AreEqual("Value", values.ElementAt(0));
+            Assert.AreEqual("Value", values.ElementAt(1));
+            Assert.AreEqual("Value", values.ElementAt(2));
+        }
+
+        [Test]
+        public async Task CanAddPoliciesAllPositions()
```

Comment:
Let's add a test that uses all 3 kinds of ways to add policies and validates that the resulting order is correc.t